src/cmd/cmd.zig

```diff
@@ -360,152 +413,126 @@ fn identity() !void {
 
 /// entrypoint to run only gossip
 fn gossip() !void {
-    var logger = try spawnLogger();
-    defer logger.deinit();
-    const metrics_thread = try spawnMetrics(logger);
-    defer metrics_thread.detach();
+    var app_base = try AppBase.init(gpa_allocator);
 
-    var exit = std.atomic.Value(bool).init(false);
-    const my_keypair = try getOrInitIdentity(gpa_allocator, logger);
-    const entrypoints = try getEntrypoints(logger);
-    defer entrypoints.deinit();
-    const my_data = try getMyDataFromIpEcho(logger, entrypoints.items);
-
-    var gossip_service = try initGossip(
-        logger,
-        my_keypair,
-        &exit,
-        entrypoints.items,
-        my_data.shred_version,
-        my_data.ip,
-        &.{},
-    );
+    var gossip_service, var gossip_manager = try startGossip(gpa_allocator, &app_base, &.{});
     defer gossip_service.deinit();
+    defer gossip_manager.deinit();
 
-    try runGossipWithConfigValues(&gossip_service);
+    gossip_manager.join();
 }
 
 /// entrypoint to run a full solana validator
 fn validator() !void {
-    var logger = try spawnLogger();
-    defer logger.deinit();
-    const metrics_thread = try spawnMetrics(logger);
-    defer metrics_thread.detach();
-
-    var rand = std.rand.DefaultPrng.init(@bitCast(std.time.timestamp()));
-    var exit = std.atomic.Value(bool).init(false);
-    const my_keypair = try getOrInitIdentity(gpa_allocator, logger);
-    const entrypoints = try getEntrypoints(logger);
-    defer entrypoints.deinit();
-    const ip_echo_data = try getMyDataFromIpEcho(logger, entrypoints.items);
+    const allocator = gpa_allocator;
+    var app_base = try AppBase.init(allocator);
 
     const repair_port: u16 = config.current.shred_collector.repair_port;
     const turbine_recv_port: u16 = config.current.shred_collector.repair_port;
 
-    // gossip
-    var gossip_service = try initGossip(
-        logger,
-        my_keypair,
-        &exit,
-        entrypoints.items,
-        ip_echo_data.shred_version, // TODO atomic owned at top level? or owned by gossip is good?
-        ip_echo_data.ip,
-        &.{
-            .{ .tag = socket_tag.REPAIR, .port = repair_port },
-            .{ .tag = socket_tag.TURBINE_RECV, .port = turbine_recv_port },
-        },
-    );
+    var gossip_service, var gossip_manager = try startGossip(allocator, &app_base, &.{
+        .{ .tag = socket_tag.REPAIR, .port = repair_port },
+        .{ .tag = socket_tag.TURBINE_RECV, .port = turbine_recv_port },
+    });
     defer gossip_service.deinit();
-    const gossip_handle = try std.Thread.spawn(.{}, runGossipWithConfigValues, .{&gossip_service});
+    defer gossip_manager.deinit();
+
+    const snapshot = try loadSnapshot(allocator, app_base.logger, gossip_service, false);
+
+    // leader schedule
+    var leader_schedule = try getLeaderScheduleFromCli(allocator) orelse
+        try leaderScheduleFromBank(allocator, &snapshot.bank);
+    const leader_provider = leader_schedule.provider();
```

Comment:
It's a form of dependency inversion. The shred collector depends on the abstract leader provider instead of a concrete implementation.

The shred collector needs to know who the leader is at a particular slot. But the the leader schedule calculation is not a part of the shred collector. So, it's not actually relevant to the shred collector *how* the leader is calculated. It just needs to be able to make a request for the slot leader and get a response.

With dependency inversion, the shred collector expresses what it needs. "I need anything that can provide me a leader for a particular slot." It doesn't impose a requirement about a particular way of calculating the leader. The shred collector is easier to run, because it doesn't ask for anything that it doesn't need.

If the shred collector depends on a particular implementation of a leader schedule calculator, it couples the shred collector with a distant region of the code. The coupling is not necessary and makes it harder to change the code later. 

I actually ran into this while implementing this code. I switched between some different approaches to calculate the leader at a particular slot. With this approach, you can make those changes without needing change the shred collector. It makes the code more modular and more easily upgraded.

Generally I strive to minimize dependencies and coupling as much as possible. I find it easier to make changes to code when the interfaces between different components are simple and do not impose unnecessary constraints.